stm32_spi: Add DMA capabilities for STM32F4 SOCs

mubes opened the pull request:
Addition of DMA for STM32F4xx family for SPI communication. To all
intents and purposes this works the same as Interrupt driven
DMA communication but is much faster and takes less CPU. Also
includes support for deferred spi start. This allows an SPI
transaction to be set up and triggered later. Useful for the
common pattern that an SPI transfer is triggered by an external
event.

Signed-off-by: Dave Marples <dave@marples.net>

erwango reviewed (commented):
Thanks for this work.
Before starting the review, can you clean up the PR a little bit:
- remove the modifications with no effect. If you feel some formatting is required, do it in a separate commit.
- split into separate commits dma and spi changes

mubes commented:
Hi Erwan,

>     * remove the modifications with no effect. If you feel some formatting is required, do it in a separate commit.

The formatting changes were demanded by checkpatch, and certainly not my idea :-(  I'll split the PullReq into substantive and prettyness commits so at least you can see where the meat is.

DAVE

mubes commented:
Erwan,

I think this is ready for a review now. I've cleaned out the changes that uncrustify wanted, and jumped through the hoops of the rest of the submission system.  I've built/tested in ProgrammedIO, Interrupt and DMA configurations cleanly.

DAVE

superna9999 commented:
@mubes Please reorganize your changes by:
- 1 commit by feature *or* fix
- use consistent commit titles by adding "spi: stm32:" for example, you can look at the git hitory to have an idea
- be sure it doesn't break other boards

We understand the needs, but we need a cleaner patchset before we accept it so it can be maintainable

mubes commented:
Neil,

OK, so the fix to the DMA handlers and the SPI DMA and delayed triggering can be considered separate features - do you want those in separate PullReqs or separate commits in a single PullReq, 'cos that's what I thought I'd just done here? If I do them as separate Pulls then it will slow down my ability to contribute these things if I have to wait for each to be approved before I can build on it.

...a bit confused by the process, but trying to understand it.

DAVE

mubes committed:
dma: stm32: Fix input validity checking order and prettyprint

Input validation checked stream direction before it had been set,
which meant it would reject valid transfers depending on what had
happened in a previous transfer. Some small prettyprint fixes too
which were done to support the investigation into this bug.

Signed-off-by: Dave Marples <dave@marples.net>

mubes committed:
spi: stm32: DMA support for STM32F4 family

Addition of DMA for STM32F4xx family for SPI communication. To all
intents and purposes this works the same as Interrupt driven
DMA communication but is much faster and takes less CPU. Kconfig
is organised so as not to break existing configurations that are
selecting interrupt based SPI.

Signed-off-by: Dave Marples <dave@marples.net>

mubes commented:
OK, PR resubmitted with three separate commits representing each functionality change (DMA fixes, SPI DMA, SPI Deferred operation).  Hopefully this addresses all of the issues.

Regards
DAVE

superna9999 commented on include/spi.h:
```diff
@@ -30,9 +30,9 @@ extern "C" {
-#define SPI_OP_MODE_MASK	0x1
+#define SPI_OP_MODE_MASTER      0
+#define SPI_OP_MODE_SLAVE       BIT(0)
+#define SPI_OP_MODE_MASK        0x1
```
These changes are not relevant

superna9999 commented on include/spi.h:
```diff
@@ -44,7 +44,7 @@ extern "C" {
  * which is the default.
  */
-#define SPI_MODE_CPOL		BIT(1)
+#define SPI_MODE_CPOL           BIT(1)
```
This change is not relevant

superna9999 commented on include/spi.h:
```diff
@@ -53,34 +53,34 @@ extern "C" {
-#define SPI_MODE_MASK		(0xE)
-#define SPI_MODE_GET(_mode_)			\
+#define SPI_MODE_MASK           (0xE)
+#define SPI_MODE_GET(_mode_) \
```
These changes are not relevant

superna9999 commented on include/spi.h:
```diff
@@ -53,34 +53,34 @@ extern "C" {
-#define SPI_TRANSFER_MSB	(0)
-#define SPI_TRANSFER_LSB	BIT(4)
+#define SPI_TRANSFER_MSB        (0)
+#define SPI_TRANSFER_LSB        BIT(4)
```
Ditto

superna9999 commented on include/spi.h:
```diff
@@ -53,34 +53,34 @@ extern "C" {
-#define SPI_WORD_SIZE_GET(_operation_)					\
+#define SPI_WORD_SIZE_SHIFT     (5)
+#define SPI_WORD_SIZE_MASK      (0x3F << SPI_WORD_SIZE_SHIFT)
+#define SPI_WORD_SIZE_GET(_operation_) \
```
Ditto

superna9999 commented on include/spi.h:
```diff
@@ -53,34 +53,34 @@ extern "C" {
 	(((_operation_) & SPI_WORD_SIZE_MASK) >> SPI_WORD_SIZE_SHIFT)
 
-#define SPI_WORD_SET(_word_size_)		\
+#define SPI_WORD_SET(_word_size_) \
```
Ditto

superna9999 commented on include/spi.h:
```diff
@@ -89,24 +89,24 @@ extern "C" {
+#define SPI_LINES_OCTAL         (3 << 11)
 
-#define SPI_LINES_MASK		(0x3 << 11)
+#define SPI_LINES_MASK          (0x3 << 11)
```
Ditto

mubes replied:
Doh! These are changes that uncrustify added. New version committed with those taken out. OK, I think we're about there now and I'm better educated as a fringe benefit.  It might be an idea to a note at https://docs.zephyrproject.org/latest/contribute/index.html to the section on uncrustify for newbies like me, suggesting to _not_ use it on existing code (or at least, make it a separate commit).

superna9999 replied:
Indeed, this is much better !

superna9999 commented on include/spi.h:
```diff
@@ -89,24 +89,24 @@ extern "C" {
  */
 /* Requests - if possible - to keep CS asserted after the transaction */
-#define SPI_HOLD_ON_CS		BIT(13)
+#define SPI_HOLD_ON_CS          BIT(13)
```
Ditto

superna9999 commented on include/spi.h:
```diff
@@ -89,24 +89,24 @@ extern "C" {
  * properly called.
  */
-#define SPI_LOCK_ON		BIT(14)
+#define SPI_LOCK_ON             BIT(14)
```
Ditto

superna9999 commented on include/spi.h:
```diff
@@ -131,9 +141,9 @@ extern "C" {
-	u32_t		delay;
+	struct device   *gpio_dev;
+	u32_t gpio_pin;
+	u32_t delay;
```
Ditto

superna9999 commented on include/spi.h:
```diff
@@ -158,9 +168,9 @@ struct spi_cs_control {
-	u16_t		slave;
+	u32_t frequency;
+	u32_t operation;
+	u16_t slave;
```
Ditto

superna9999 reviewed (commented):
Much better! make sure you don't change formatting except the feature you are adding

mubes committed:
spi: stm32: Deferred triggering mode for STM32F4

Support for deferred spi start on STM32F4. This allows an SPI
transaction to be set up and triggered later. Useful for the
common pattern that an SPI transfer is triggered by an external
event. Does not impact normal operation when not enabled.

Signed-off-by: Dave Marples <dave@marples.net>

avisconti commented:
Good job.
I have few ways to test it out (one is 96b_argonkey/sensor). I will do it tomorrow.

mubes committed:
dma: stm32: Tidying of interrupt enables

Tidy and simplify interrupt enables so that interrupts are not
permanently enabled for all DMA streams on both controllers but
are instead turned on and off as needed.

Signed-off-by: Dave Marples <dave@marples.net>

mubes committed:
spi: stm32: Documentation update

Fixes to documentation to reflect use of the defer_mode bit in the
operation structure.

Signed-off-by: Dave Marples <dave@marples.net>

erwango commented:
@tbursztyka , when you have some time, this PR has impacts on spi.h

tbursztyka commented on include/spi.h:
```diff
@@ -218,6 +228,13 @@ typedef int (*spi_api_io_async)(struct device *dev,
+ * @brief Callback API for triggering SPI device.
+ * See spi_trigger() for argument descriptions
+ */
+typedef int (*spi_api_trigger)(struct device *dev,
```
What would be the uses cases for this?

What would require one to pre-load a spi transaction, and wait for a signal to actually start it, instead of waiting for that signal to do the transaction as usual?

tbursztyka reviewed (changes requested):
I am doubtful about this new trigger operation. We already have many drivers that wait for a signal to start a spi transaction and it works already with existing API.

I did not go in the details, but there are obvious mistakes like breaking spi_config struct alignment or forgetting to lock the spi controller for a trigger to happen.

Note that besides this trigger stuff, rest looks fine to me.

mubes commented:
> I am doubtful about this new trigger operation. We already have many drivers that wait for a signal to start a spi transaction and it works already with existing API.

I hoped the scope shots in my original post would have provided the use case, but here's some more information to support it.  Firstly, a scope shot of a transaction done with a 'conventional' trigger mechanism (semaphore triggered).  It's possible there are slightly faster OS supported mechanisms such as polls to do this, but the principle holds.  In this shot the yellow trace falling edge is the trigger and the red trace is the SPI clock for a 24-octet transfer);

![semaphore](https://user-images.githubusercontent.com/5451919/52716450-b16fba00-2f96-11e9-96b4-c7f15059aa08.png)

So, in this case the transfer starts after 26.2uS and completes after 45.5uS.  Per my original notes above this transfer previously took around 116uS in interrupt mode. Now, if we use the deferred transfer mechanism, which is a trival change to user code (replace the` k_sem_give` with `spi_trigger` and OR `SPI_DEFER_TRANSFER` onto the `config.operation` member) we get;

![defer](https://user-images.githubusercontent.com/5451919/52716682-38bd2d80-2f97-11e9-9eea-46a129784d38.png)

In this shot the orange reference trace is the previous, semaphore-triggered, transfer. This time the transfer is underway after 3.4uS and completes in 22.7uS - before the semaphore-triggered case has even started!

The justification is therefore straightforward; A nearly 8-fold reduction in transfer startup latency and, overall, faster completing transfers.

This behaviour is essential in lots of circumstances. A trivial example would be using the SPI to output video (see https://www.youtube.com/watch?v=5UFpp3ao460&feature=youtu.be for the kind of thing I mean). My real-world use case is 24-octet transfers which occur every 100uS which can't afford a 45.5uS transfer time. There are lots of use cases where there is spare time available _after_ the transfer is complete to prepare the next one, but no opportunity for latency when it is requested....I'm afraid the fact that it isn't _always_ necessary isn't a good defence against the argument that is _sometimes_ necessary and if Zephyr is to meet it's objective of being a widely adopted OS its these kinds of edge use-cases that will make the difference.

> I did not go in the details, but there are obvious mistakes like breaking spi_config struct alignment or forgetting to lock the spi controller for a trigger to happen.

OK, these I will need advice on, but I'm sure they can be fixed with the right help.

Regards
DAVE

tbursztyka commented:
@mubes ok, now it's clearer. Always expose use-cases for such change.

> OK, these I will need advice on, but I'm sure they can be fixed with the right help.

Hopefully it's quite trivial. Although you'll probably need a dedicated PR for this trigger feature.

- First, split your patch: have one for spi.h, and another one for your driver. (Actually, I'd like to see other drivers being updated as well)
- Revise the struct spi_config, you are breaking the alignment. (move the slave attribute after the cs attribute. Size is going to be the same, but at least if there is additional stuff there someday it won't be a mess)
- Now, about locking the spi controller: since the transaction is deferred, you need to avoid that another call to the same spi controller messes up your transaction: you need to make it wait the controller becomes available again. For this, there is a lock_on bit in operation: rule would be that if one want to use defer, he has to set lock_on as well (or then it would be -EINVAL I guess)
- last one, merge your last commit into this spi.h trigger commit.

Note that we are on feature freeze for 1.14 afaik, so this will have to wait that new release cycle starts before it can get in.

erwango commented:
@mubes, do you think about completing this proposal or should  we close it?

erwango commented:
PR stalled since almost one year.
Would require significant rework to fit new DMA representation (using dts).
Closing
